From 2e7042d9551aa01663003a9da4ba109dc242ea56 Mon Sep 17 00:00:00 2001
From: Leonard Crestez <leonard.crestez@nxp.com>
Date: Thu, 3 May 2018 17:52:09 +0300
Subject: [PATCH 3667/5242] MLK-18138-2: soc: imx: gpc: Convert imx6sx to new
 bindings

commit  cbe87526614200ad723cdad08d77c472dacf324f from
https://source.codeaurora.org/external/imx/linux-imx.git

This fixes graphics on imx6sx by aligning closer to upstream instead of
adding new features to old bindings.

Upstream adds a 4th power domain for PCI but this is is wrong: the PCI
block is in the DISPMIX domain and only PCIE_PHY is in the PCIE_PHY
power domain.

Manual is not very clear on this but in section 10.4.1.4.1 there is this
statement: "The DISPLAY domain contains GIS, CSI, PXP, LCDIF, PCIe,
DCIC, and LDB.  It is supplied by internal regulator."

Placing pcie in a 4th power domain makes lspci hang when display is
turned off.

In upstream the dispmix domain is not actually touched on 6sx so it's
always on, this is why pci seems to work.

Signed-off-by: Leonard Crestez <leonard.crestez@nxp.com>
Acked-by: Richard Zhu <hongxing.zhu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/boot/dts/imx6sx.dtsi |   32 +++++++++++++++++++-------------
 1 file changed, 19 insertions(+), 13 deletions(-)

diff --git a/arch/arm/boot/dts/imx6sx.dtsi b/arch/arm/boot/dts/imx6sx.dtsi
index 6a3468c..8434d0b 100644
--- a/arch/arm/boot/dts/imx6sx.dtsi
+++ b/arch/arm/boot/dts/imx6sx.dtsi
@@ -303,7 +303,7 @@
 				"gpu3d_shader_clk";
 			resets = <&src 0>;
 			reset-names = "gpu3d";
-			power-domains = <&gpc 1>;
+			power-domains = <&pd_pu>;
 		};
 
 		gpmi: gpmi-nand@1806000{
@@ -897,15 +897,9 @@
 				interrupts = <GIC_SPI 89 IRQ_TYPE_LEVEL_HIGH>;
 				interrupt-parent = <&intc>;
 				fsl,mf-mix-wakeup-irq = <0x7c00000 0x3d00 0x0 0x400240>;
-				clocks = <&clks IMX6SX_CLK_GPU>, <&clks IMX6SX_CLK_IPG>,
-					<&clks IMX6SX_CLK_PXP_AXI>, <&clks IMX6SX_CLK_DISPLAY_AXI>,
-					<&clks IMX6SX_CLK_LCDIF1_PIX>, <&clks IMX6SX_CLK_LCDIF_APB>,
-					<&clks IMX6SX_CLK_LCDIF2_PIX>, <&clks IMX6SX_CLK_CSI>,
-					<&clks IMX6SX_CLK_VADC>;
-				clock-names = "gpu3d_core", "ipg", "pxp_axi", "disp_axi", "lcdif1_pix",
-						"lcdif_axi", "lcdif2_pix", "csi_mclk";
+				clocks = <&clks IMX6SX_CLK_IPG>;
+				clock-names = "ipg";
 				pcie-phy-supply = <&reg_pcie_phy>;
-				#power-domain-cells = <1>;
 
 				pgc {
 					#address-cells = <1>;
@@ -928,6 +922,18 @@
 						#power-domain-cells = <0>;
 						power-supply = <&reg_pcie>;
 					};
+
+					pd_disp: power-domain@2 {
+						reg = <2>;
+						#power-domain-cells = <0>;
+						clocks = <&clks IMX6SX_CLK_PXP_AXI>,
+							 <&clks IMX6SX_CLK_DISPLAY_AXI>,
+							 <&clks IMX6SX_CLK_LCDIF1_PIX>,
+							 <&clks IMX6SX_CLK_LCDIF_APB>,
+							 <&clks IMX6SX_CLK_LCDIF2_PIX>,
+							 <&clks IMX6SX_CLK_CSI>,
+							 <&clks IMX6SX_CLK_VADC>;
+					};
 				};
 			};
 
@@ -1426,7 +1432,7 @@
 					clocks = <&clks IMX6SX_CLK_PXP_AXI>,
 						 <&clks IMX6SX_CLK_DISPLAY_AXI>;
 					clock-names = "pxp-axi", "disp-axi";
-					power-domains = <&gpc 2>;
+					power-domains = <&pd_disp>;
 					status = "disabled";
 				};
 
@@ -1449,7 +1455,7 @@
 						 <&clks IMX6SX_CLK_LCDIF_APB>,
 						 <&clks IMX6SX_CLK_DISPLAY_AXI>;
 					clock-names = "pix", "axi", "disp_axi";
-					power-domains = <&gpc 2>;
+					power-domains = <&pd_disp>;
 					status = "disabled";
 				};
 
@@ -1461,7 +1467,7 @@
 						 <&clks IMX6SX_CLK_LCDIF_APB>,
 						 <&clks IMX6SX_CLK_DISPLAY_AXI>;
 					clock-names = "pix", "axi", "disp_axi";
-					power-domains = <&gpc 2>;
+					power-domains = <&pd_disp>;
 					status = "disabled";
 				};
 
@@ -1472,7 +1478,7 @@
 					clocks = <&clks IMX6SX_CLK_VADC>,
 						 <&clks IMX6SX_CLK_CSI>;
 					clock-names = "vadc", "csi";
-					power-domains = <&gpc 2>;
+					power-domains = <&pd_disp>;
 					gpr = <&gpr>;
 					status = "disabled";
 				};
-- 
1.7.9.5

